tests/models/hf/test_fsdp_weight_tying.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2024 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+import pathlib
+from typing import Optional
+
+import pytest
+from composer import Trainer
+from omegaconf import OmegaConf as om
+
+from llmfoundry import COMPOSER_MODEL_REGISTRY
+from llmfoundry.utils.builders import build_tokenizer
+
+
+@pytest.mark.world_size(2)
+@pytest.mark.gpu
+@pytest.mark.parametrize('peft_config', [
+    None, {
+        'peft_type': 'LORA',
+        'task_type': 'CAUSAL_LM',
+        'lora_alpha': 32,
+        'lora_dropout': 0.05,
+        'r': 16,
+        'target_modules': [
+            'q_proj',
+            'k_proj',
+            'v_proj',
+        ],
+    }
+])
+@pytest.mark.parametrize('init_device', ['cpu', 'mixed', 'meta'])
+def test_fsdp_weight_tying(peft_config: Optional[dict], tmp_path: pathlib.Path,
+                           init_device: str):
+    model_cfg = {
+        'name': 'hf_causal_lm',
+        'pretrained_model_name_or_path': 'mistralai/Mistral-7B-v0.1',
+        'config_overrides': {
+            'num_hidden_layers': 2,
+            'hidden_size': 32,
+            'intermediate_size': 64,
+            'tie_word_embeddings': True,
+        },
+        'use_auth_token': True,
+        'pretrained': False,
+        'init_device': init_device,
+    }
+    tokenizer_name = 'mistralai/Mistral-7B-v0.1'
+
+    assert model_cfg is not None
+    assert tokenizer_name is not None
+    model_cfg = om.create(model_cfg)
+    if peft_config is not None:
+        model_cfg['peft_config'] = peft_config
+
+    fsdp_config = {
+        'sharding_strategy': 'FULL_SHARD',
+        'mixed_precision': 'PURE',
+        'activation_checkpointing': False,
+        'activation_checkpointing_reentrant': False,
+        'activation_cpu_offload': False,
+        'limit_all_gathers': True,
+        'state_dict_type': 'full',
+        'sync_module_states': True,
+    }
+
+    tokenizer = build_tokenizer(
+        tokenizer_name=tokenizer_name,
+        tokenizer_kwargs={'model_max_length': 32},
+    )
+
+    original_model = COMPOSER_MODEL_REGISTRY[model_cfg['name']](model_cfg,
+                                                                tokenizer)
+
+    lm_head = original_model.model.lm_head if peft_config is None else original_model.model.base_model.model.lm_head
```

Comment:
use a common var for `underlying_model`

```
underlying_model = original_model.model if peft_config is None else original_model.model.base_model.model

```

Reply:
can we use `maybe_get_underlying_model` from composer after the release?

Reply:
ah yeah, should be able to update to use that helper